pageserver/src/storage_sync/delete.rs

```diff
@@ -1,230 +1,48 @@
 //! Timeline synchronization logic to delete a bulk of timeline's remote files from the remote storage.
 
-use anyhow::Context;
-use futures::stream::{FuturesUnordered, StreamExt};
-use tracing::{debug, error, info};
+use anyhow::{Context, Result};
+use tracing::debug;
 
-use crate::storage_sync::{SyncQueue, SyncTask};
-use remote_storage::RemoteStorage;
-use utils::zid::ZTenantTimelineId;
+use std::path::Path;
 
-use super::{LayersDeletion, SyncData};
+use remote_storage::GenericRemoteStorage;
+use remote_storage::RemoteStorage;
 
 /// Attempts to remove the timleline layers from the remote storage.
 /// If the task had not adjusted the metadata before, the deletion will fail.
-pub(super) async fn delete_timeline_layers<'a, P, S>(
-    storage: &'a S,
-    sync_queue: &SyncQueue,
-    sync_id: ZTenantTimelineId,
-    mut delete_data: SyncData<LayersDeletion>,
-) -> bool
+pub(super) async fn delete_layer(
+    storage: &GenericRemoteStorage,
+    local_layer_path: &Path,
+) -> Result<()> {
+    match storage {
+        GenericRemoteStorage::Local(s) => delete_layer_impl(s, local_layer_path).await,
+        GenericRemoteStorage::S3(s) => delete_layer_impl(s, local_layer_path).await,
+    }
+}
```

Comment:
BTW, this is a reoccurring pattern with GenericRemoteStorage, even before this PR. I'd like to refactor GenericRemoteStorage so that you could call its functions directly, without always doing this match+generic function trick. Perhaps GenericRemoteStorage could be just `dyn RemoteStorage`.

That's not possible today because of the RemoteObjectId type that's part of the RemoteStorage trait and is used in its functions. How about replacing RemoteObjectId with just `String`? Or if we want to have type-safety to make sure we don't accidentally mix up local and remote paths, it could be newtype over String.

@SomeoneToIgnore or @LizardWizzard want to pick that up?

Reply:
`GenericRemoteStorage` was created to ease passing it between modules, otherwise it would requre a few generics specified.

Yet in the place where it's supposed to be used, I don't see an issue writing `storage: &S` , the generics 
https://github.com/neondatabase/neon/blob/f38f45b01d4fdc3aa6b682f98a2ccad571b79a2b/pageserver/src/storage_sync.rs#L1416-L1418 and call the methods directly.

I'm not fond of the `String` as a replacement, since S3 flat path is different from local fs `PathBuf` and other storages might something else.
A newtype might work, but again, not sure how that scales with more storages added.